### PR TITLE
CI Visibility support in Agentless mode

### PR DESCRIPTION
**Pull request by albertvaka**
### What does this PR do?

CI Visibility was only available when running with an Agent, but not for Agentless. This allows CI Visibility to be enabled in Agentless mode.

### Description of the Change

Implements the `startBuildTrace`, `finishBuildTrace` and `sendPipelineTrace` in `DatadogHttpClient` which previously were only implemented in the `DatadogAgentClient`. 

For Agentless reporting we use CI Visibility Webhooks API, which unlike the Agent API doesn't receive traces. Because of this I had to add `DatadogWebhookBuildLogic` and `DatadogWebhookPipelineLogic` to generate payloads for the Webhooks API, equivalent to the existing `DatadogBuildPipelineLogic` and `DatadogTracePipelineLogic` that generate spans.

### Known issues

These two "features" were working when using APM through the Agent and won't work in Agentless since the Webhooks API still doesn't support them:
* ~Nested stages (stages that have another stage as their parent)~ (fixed)
* Stage breakdown graph in Datadog

### Verification Process

Run different kinds of jobs and pipelines with the two modes and compare the data on the Datadog app.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildWebhookAction.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import hudson.model.InvisibleAction;
+import org.datadog.jenkins.plugins.datadog.model.BuildData;
+
+import java.io.Serializable;
+
+/**
+ * Keeps build bebhook propagation
+ */
+public class BuildWebhookAction extends InvisibleAction implements Serializable {
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildWebhookAction.java:11]: Have you added this action in the `cleanUpTraceActions` method?

**Review by drodriguezhdez — COMMENTED** [review]
I dropped a couple of comments. Probably some refactor may be worth it. There are several classes that look pretty similar to each other.

File: src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java [hunk 2]
```diff
@@ -61,6 +61,7 @@ public String getBuildLevel() {
     private final List<BuildPipelineNode> children;
     private final String id;
     private final String name;
+    private String stageId;
```
[the remark this answers is not in the document]
Comment (albertvaka) [reply src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildPipelineNode.java:64]: The jobs need to specify their parent stage ID in the Webhooks API.

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java [hunk 3]
```diff
@@ -554,6 +590,78 @@ private boolean postLogs(String url, Secret apiKey, String payload) {
...
     }
 
+    /**
+     * Posts a given payload to the Datadog Webhook Intake, using the user configured apiKey.
+     *
+     * @param payload - A webhooks payload.
+     * @return a boolean to signify the success or failure of the HTTP POST request.
+     */
+    @SuppressFBWarnings("REC_CATCH_EXCEPTION")
+    public boolean postWebhook(String payload) {
```
[the remark this answers is not in the document]
Comment (albertvaka) [reply src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java:600]: I tried but there are subtle differences. We might be able to get rid of them, but I don't want to risk breaking something so I would do it on a separate PR.

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java [hunk 4]
```diff
@@ -782,6 +783,26 @@ public static String getNormalizedResultForTraces(@Nonnull String result) {
         }
     }
 
+    /**
+     * Returns a normalize result for traces.
```
Comment (sarah-witt) [inline src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:787]:
```suggestion
     * Returns a normalized result for traces.
```
nit

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java [hunk 5]
```diff
@@ -782,6 +783,26 @@ public static String getNormalizedResultForTraces(@Nonnull String result) {
...
+     * @param result (success, failure, error, aborted, not_build, canceled, skipped, unknown)
+     * @return the normalized result for the traces based on the jenkins result
+     */
+    public static String statusFromResult(@Nonnull String result) {
```
Comment (sarah-witt) [inline src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:791]: Why is this different from `getNormalizedResultForTraces`?

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java [hunk 6]
```diff
@@ -793,7 +814,7 @@ public static void severe(Logger logger, Throwable e, String message){
         if(e != null) {
             StringWriter sw = new StringWriter();
             e.printStackTrace(new PrintWriter(sw));
-            logger.finer(message + ": " + sw.toString());
```
Comment (sarah-witt) [inline src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:796]: This change should be a good addition but we might want to do some testing when there are severe errors. We might want to change this to info or debug.

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java [hunk 7]
```diff
@@ -622,20 +770,50 @@ private String getJenkinsVersion(){
 
     @Override
     public boolean startBuildTrace(BuildData buildData, Run run) {
-        logger.fine("Traces are only available using Datadog Agent.");
-        return false;
+        if(this.isWebhookIntakeConnectionBroken()){
+            logger.severe("Your client is not initialized properly");
```
Comment (sarah-witt) [inline src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java:774]:
```suggestion
            logger.severe("Unable to start build trace; your client is not initialized properly.");
```
nit to make debugging easier

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java [hunk 8]
```diff
@@ -622,20 +770,50 @@ private String getJenkinsVersion(){
...
+            return false;
+        }
     }
 
     @Override
     public boolean finishBuildTrace(BuildData buildData, Run<?, ?> run) {
-        logger.fine("Traces are only available using Datadog Agent.");
-        return false;
+        if(this.isWebhookIntakeConnectionBroken()){
+            logger.severe("Your client is not initialized properly");
```
Comment (sarah-witt) [inline src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java:790]:
```suggestion
            logger.severe("Unable to finish build trace; your client is not initialized properly.");
```

**Review by sarah-witt — COMMENTED** [review]
Looks great 🎉 Left a few nits! Can you also update the `README.md` describing the new environment variables/settings, as well as make a note that CI visibility is available for both methods?

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java [hunk 9]
```diff
@@ -160,7 +169,19 @@ public void validateConfiguration() throws IllegalArgumentException {
...
         if (DatadogUtilities.getDatadogGlobalDescriptor().getEnableCiVisibility() ) {
-            logger.warning("Traces Collection only can be used if Datadog Agent reports to Datadog.");
+            if (webhookIntakeUrl == null || webhookIntakeUrl.isEmpty()) {
+                throw new IllegalArgumentException("Datadog Webhook Intake URL is not set properly");
+            }
+            try {
+                boolean webhookConnection = validateWebhookIntakeConnection(webhookIntakeUrl, apiKey);
+                if (!webhookConnection) {
+                    instance.setWebhookIntakeConnectionBroken(true);
+                    logger.warning("Connection broken, please double check both your Webhook Intake URL and Key");
```
Comment (sarah-witt) [inline src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java:179]: I see we also do this for the log intake connection, but do you mind making the log lines different so it's easier to debug?

**Review by sarah-witt — COMMENTED** [review]
Just a few more small comments, otherwise looks good!

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java [hunk 10]
```diff
@@ -554,6 +590,78 @@ private boolean postLogs(String url, Secret apiKey, String payload) {
...
+        logger.fine("Sending webhook");
+        if(this.isWebhookIntakeConnectionBroken()){
+            logger.severe("Your client is not initialized properly");
+            return false;
+        }
+
+        HttpURLConnection conn = null;
+        try {
+            logger.fine("Setting up HttpURLConnection...");
+            String urlParameters = "?service=" + DatadogUtilities.getDatadogGlobalDescriptor().getCiInstanceName();
```
[the remark this answers is not in the document]
Comment (albertvaka) [reply src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java:613]: I've refactored the duplicated code into a separate method in 54c3da0459c459c117b0e7e9d13ef47e74b7aa5a

**Review by sarah-witt — APPROVED** [review]

**Pull request merged**